### ISSUE TITLE
Globs ran in the init phase causing increased build time on all commands

### ISSUE THREAD

**Issue by filipesilva** [bazelbuild/rules_go#1862]
Heya,

While profiling https://github.com/angular/angular bazel build times, we noticed that a long time (~2.5s) was being spent in the init phase even when the entire build was cached. Most of it seemed to be `VFS_DIR` calls.

```
=== INIT PHASE INFORMATION ===

Total init phase time                    2.480 s

Total time (across all threads) spent on:
              Type    Total    Count     Average
              INFO    2.50%       67     36.8 ms
          VFS_STAT    0.00%        1     0.30 ms
           VFS_DIR   97.48%     1640     58.6 ms
        VFS_DELETE    0.00%        2     0.35 ms
          VFS_OPEN    0.00%        2     0.33 ms
          VFS_READ    0.00%        2     0.09 ms
         VFS_WRITE    0.00%        3     0.08 ms
```

We dug into it in https://github.com/angular/angular/issues/27427 and our best bet for the cause was a couple of globs in templates by both `rules_nodejs` and `rules_go`. We had seen the effect of these globs on templates before (https://github.com/bazelbuild/rules_nodejs/pull/405) so were fairly confident it was related.

Replacing those globs by static file lists in https://github.com/bazelbuild/rules_nodejs/pull/472 reduced the VFS_DIR calls by roughly half, together with lowe init phase time (2.48s -> 2s). The VFS_DIR calls to `external/nodejs` were gone but the ones to `external/go_sdk` remain.
```
=== INIT PHASE INFORMATION ===

Total init phase time                    2.008 s

Total time (across all threads) spent on:
              Type    Total    Count     Average
              INFO    2.38%       67     29.9 ms
          VFS_STAT    0.00%        1     0.20 ms
           VFS_DIR   97.61%      854     96.0 ms
        VFS_DELETE    0.00%        2     0.20 ms
          VFS_OPEN    0.00%        1     0.29 ms
         VFS_WRITE    0.00%        3     0.05 ms
```

I prepared a basic repro where you can see these calls in https://github.com/filipesilva/rules_go_init_phase:
```
git clone https://github.com/filipesilva/rules_go_init_phase
cd rules_go_init_phase
bazel run :binary
bazel run :binary --profile debug.profile --record_full_profiler_data
bazel analyze-profile debug.profile 
bazel analyze-profile debug.profile --task_tree ".*" --task_tree_threshold 0 > task-tree-log.txt
```

The `bazel analyze-profile debug.profile` command should show you similar profiling information as below, where 840 VFS_DIR calls are listed:
```
=== INIT PHASE INFORMATION ===

Total init phase time                     270 ms

Total time (across all threads) spent on:
              Type    Total    Count     Average
              INFO   24.87%       67     3.91 ms
          VFS_STAT    0.02%        1     0.19 ms
           VFS_DIR   74.35%      840     0.93 ms
        VFS_DELETE    0.03%        2     0.17 ms
          VFS_OPEN    0.03%        1     0.32 ms
         VFS_WRITE    0.02%        3     0.06 ms
```

The `bazel analyze-profile debug.profile --task_tree ".*" --task_tree_threshold 0 > task-tree-log.txt` command will generate a log where you can see all these calls. I've also committed the file if you want to see it without running the repro: https://github.com/filipesilva/rules_go_init_phase/blob/master/task-tree-log.txt

It's worthwhile to mention that in the original repro the average VFS_DIR length is about 10x longer. I don't know why, but suspect it's related with high concurrent disk IO related to other disk operations happening at the same time and not reported due to the large repo, but don't really know.

The file we suspect has the globs is 
https://github.com/bazelbuild/rules_go/blob/368049409d12d0827dcfd0bd401188c22651cc29/go/private/BUILD.sdk.bazel#L57-L64

cc @meteorcloudy

**Comment by jayconrod** [comment]
Very interesting. Thanks for investigating this.

I imagine replacing this glob with something equivalent in the SDK repository rules would be straightforward.

I wonder if this should be an item in Bazel's [Optimizing Performance](https://docs.bazel.build/versions/master/skylark/performance.html) document?

**Comment by meteorcloudy** [comment]
I think it would be a good item to add to [Optimizing Performance](https://docs.bazel.build/versions/master/skylark/performance.html).
/cc @laurentlb @meisterT

**Comment by jayconrod** [comment]
Closing this issue. I implemented an optimization (#1884). It reduced VFS_DIR operations but wasn't actually faster.

**Comment by filipesilva** [comment]
Thanks for taking a stab at it @jayconrod. I'll ping again if I find another multiple-second init phase case, and will try to get a better repro to illustrate it then.
